homeassistant/components/recorder/run_history.py

```diff
@@ -0,0 +1,144 @@
+"""Track recorder run history."""
+from __future__ import annotations
+
+import bisect
+from dataclasses import dataclass
+from datetime import datetime
+
+from sqlalchemy.orm.session import Session
+
+import homeassistant.util.dt as dt_util
+
+from .models import RecorderRuns, process_timestamp
+
+
+@dataclass
+class _RecorderRunsHistory:
+    """Bisectable history of RecorderRuns."""
+
+    run_timestamps: list[int]
+    runs_by_timestamp: dict[int, RecorderRuns]
+
+
+class RunHistory:
+    """Track recorder run history."""
+
+    def __init__(self) -> None:
+        """Track recorder run history."""
+        self._recording_start = dt_util.utcnow()
+        self._current_run_info: RecorderRuns | None = None
+        self._run_history = _RecorderRunsHistory([], {})
+
+    @property
+    def recording_start(self) -> datetime:
+        """Return the time the recorder started recording states.
+
+        This function is thread safe.
+        """
+        return self._recording_start
+
+    @property
+    def current(self) -> RecorderRuns:
+        """Get the current run.
+
+        This function is thread safe.
+        """
+        assert self._current_run_info is not None
+        return self._current_run_info
+
+    def get(self, start: datetime) -> RecorderRuns | None:
+        """Return the recorder run that started before or at start.
+
+        If the first run started after the start, return None
+
+        This function is thread safe.
+        """
+        if start >= self.recording_start:
+            return self.current
+
+        # Hold a reference here to self._run_history
+        # to ensure it does not get changed out from
+        # under us in the recorder thread.
+        run_history = self._run_history
+
+        # Ensure we do not access self._run_history
+        # after this point as the other thread
+        # could change it out on us
```

Comment:
This comment seems confusing, we've already made a copy of `self._run_history`. Can it be rephrased?

Reply:
Removed as the comment is no longer needed due to refactoring